src/simulator.zig

```diff
@@ -457,6 +459,49 @@ pub const Simulator = struct {
         return false;
     }
 
+    // Returns a header for a prepare which can't be repaired by the core due to storage faults.
+    //
+    // When generating a FaultAtlas, we don't try to protect core from excessive errors. Instead,
+    // if the core gets stuck, we verify that this is indeed due to storage faults.
+    pub fn core_missing_prepare(simulator: *const Simulator) ?vsr.Header {
+        assert(simulator.core.count() > 0);
+
+        var missing_op: ?u64 = null;
+        for (simulator.cluster.replicas) |replica| {
+            if (simulator.core.isSet(replica.replica)) {
+                assert(simulator.cluster.replica_health[replica.replica] == .up);
+                assert(replica.status == .normal);
+                if (replica.commit_min < replica.op) {
+                    if (missing_op == null or missing_op.? > replica.commit_min + 1) {
+                        missing_op = replica.commit_min + 1;
+                    }
+                }
+            }
+        }
+
+        if (missing_op == null) return null;
+
+        const missing_header = simulator.cluster.state_checker.header_with_op(missing_op.?);
+
+        for (simulator.cluster.replicas) |replica| {
+            if (simulator.core.isSet(replica.replica) and
+                replica.replica < simulator.cluster.replica_count)
```

Comment:
Huh, forgot that we have `.standby`...

This is actually something I am struggling with respect to our code style, what's our stance wrt [the law of demeter](https://www.tedinski.com/2018/12/18/the-law-of-demeter.html)? If I want to peek through several abstraction layers, should I just go and reach inside through a layer of fields, or should I introduce a helper function?

In the `has_clean` case, I copied the code from `on_request_prepare`, which open-coded the logic for "we might have a specific prepare, even if we are not sure about" in `replica.zig`, and I missed that we have similar (and better) logic in journal.zig.

Reply:
>When writing tests I am fine doing whichever is simpler (which is usually just reaching through layers of fields).

Yeah, same here, though for me rational is not so much simplicity, as the desire to keep production code as simple as possible (eg, rather than buggify _in the source_, we prefer to buggify in the simulator, to keep the source pristine).